### ISSUE TITLE
Retain joinpoint locations on Clava.rebuild()

### ISSUE THREAD

**Issue by Sleepy105** [specs-feup/clava#41]
Wrap up of the discussion:

- Maintain a record of all joinpoint objects currently in existence (vars in a LARA/JS script).
- When doing a rebuild, target each joinpoint on record with a distinctive pragma.
- Pass the annotated code through Clang.
- Map the new joinpoints to the existing variables.
- Remove the pragmas annotations.
- Return to user space.

Problems:

1. Non-code joinpoints (file, type, etc.) can't be annotated with pragmas.

**Comment by joaobispo** [comment]
Join points are wrappers around AST nodes that provide a common interface and allow different languages with different ASTs to be handled in a similar way.

Taking this into consideration, join points are treated as shorted-lived objects, every time we access an AST node, a new join point is potentially created. Persistence should be viewed at the level of the AST, hence the current decision to store data objects as pragmas, whenever a node supports a pragma, and data objects for other nodes are mapped to AST node ids, not join points.

When a join point accesses a data object, it uses a map of data objects for the current context. Pushing and popping ASTs from the AST stack do not change the context of data objects, so if a node maintains its id, it is possible to communicate information between pushes and pops.

Since node ids do not persist between rebuilds, we only persist data objects of nodes that can be directly annotated with pragmas.


**In short**: all data objects are mapped by node id (i.e. `$jp.astId`), and persist across `Clava.pushAst()/popAst()`. Only nodes that support pragma annotations (e.g. `$statement`, `$function`) persist data objects across `Clava.rebuild()`.
